PassiveTotal service

ms-xy opened the pull request:
Query PassiveTotal.org for enrichment data to Domains, IPs, Emails, or Indicators.

ms-xy committed:
PassiveTotal service.

Query PassiveTotal.org for enrichment data to Domains, IPs, Emails, or Indicators.

Signed-off-by: Maximilian Schott <ms@roottec.com>

ms-xy committed:
Timeout is now actually obeyed.

Signed-off-by: Maximilian Schott <ms@roottec.com>

webstergd commented on config/docker-compose.yml.example:
```diff
@@ -26,42 +26,50 @@ objdump:
+passivetotal:
+  build: ../src/main/scala/org/holmesprocessing/totem/services/passivetotal
+  ports:
+    - "7720:8080"
```
set port to 9710 please. Gives us space and I set all the no files to 97xx in a previous pull request

webstergd commented on config/totem.conf.example:
```diff
@@ -62,24 +62,28 @@ totem {
       resultRoutingKey = "objdump.result.static.totem"
     }
+    passivetotal {
+      uri = ["http://127.0.0.1:7720/passivetotal/"]
```
same as above

webstergd commented on config/totem.conf.example:
```diff
@@ -62,24 +62,28 @@ totem {
+    }
     peid {
-      uri = ["http://127.0.0.1:7720/peid/"]
+      uri = ["http://127.0.0.1:7730/peid/"]
```
revert the ports please as it should be 97xx and wouldn't require these to change.

webstergd commented:
Lets code review this. I would like to learn more about what you did and why. should be good for the other to learn as well.

ms-xy commented:
Didn't realize that you changed the ports and that we now have special port ranges for specific services. (It's still alphabetical, so I inserted it where it would have gone if it was still purely alphabetical as it was before).
I'll adjust that in a moment.

Edit: regarding code-review, just tell me when and I'll be ready.

ms-xy committed:
Port number fix as requested.

Signed-off-by: Maximilian Schott <ms@roottec.com>

ms-xy committed:
Updated branch.

Signed-off-by: Maximilian Schott <ms@roottec.com>

webstergd commented:
yeah with GSoC finishing up this weekend, we had lots and lots of pull requests.

ms-xy committed:
Adjusted configs / Dockerfile to new format. Added support for config download as it now exists for the other services.

Signed-off-by: Maximilian Schott <ms@roottec.com>

webstergd commented on src/main/scala/org/holmesprocessing/totem/services/passivetotal2/LICENSE:
```diff
@@ -0,0 +1,16 @@
+limitations under the License.
+
+# Implementation for TOTEM made by Maximilian Schott
+# Objdump is available under the GPL (https://www.gnu.org/licenses/gpl.html)
```
remove objdump

ms-xy replied:
I honestly forgot I added that there for objdump :/ my bad, good catch

webstergd commented on src/main/scala/org/holmesprocessing/totem/services/passivetotal2/PassiveTotalREST.scala:
```diff
@@ -0,0 +1,37 @@
+      .map({
+      case Right(content) =>
+        PassiveTotalSuccess(true, JString(content), Arguments)
+
```
why did we drop the check for 400, 404, and 500?

webstergd commented on src/main/scala/org/holmesprocessing/totem/services/passivetotal2/README.md:
```diff
@@ -0,0 +1,17 @@
+Since this container needs to have access to the sample file,
+you need to run this container with:
+
+`-v /tmp:/tmp:ro`
```
remove /tmp:/tmp:ro as it doesn't need to access tmp

ms-xy committed:
Fixes as requested.

Signed-off-by: Maximilian Schott <ms@roottec.com>

ms-xy commented:
Lets discuss what to do about the error codes in the code-review, I'm not entirely certain how to handle request errors, as we don't have partial result sets in Totem (?).

cynexit commented on src/main/scala/org/holmesprocessing/totem/services/passivetotal2/src/passivetotal-service/passivetotal/queryobject/query-object.go:
```diff
@@ -0,0 +1,89 @@
+// All other types are resorted to being an Indicator.
+func (self *Object) detectObjectType() {
+
+    if net.ParseIP(self.Obj) != nil {
```
Check this again

ms-xy committed:
Fixes and improvements.

+ Switched to upstream jason library, they accepted the pull request
+ Improved build process
+ Added detailed errors
+ Improved and modified object type detection mechanism
+ Enabled overriding of query settings on per-request basis
- Disabled SSL information query (currently missing a way to determine if object is applicable or not)
+ Moved base URL for queries to constant for easier maintenance
+ Reformatted query functions for better readability

Signed-off-by: Maximilian Schott <ms@roottec.com>

ms-xy committed:
Merge remote-tracking branch 'upstream/master' into passivetotal-service-pull

Signed-off-by: Maximilian Schott <ms@roottec.com>

ms-xy committed:
Small cosmetic change + additional error code.

Signed-off-by: Maximilian Schott <ms@roottec.com>

ms-xy commented:
Updated the pull request, added error codes as requested, improved the object type detection mechanism. Some other small fixes and cosmetic changes too.
Took out SSL API query as discussed with @webstergd .

webstergd commented on src/main/scala/org/holmesprocessing/totem/services/passivetotal2/Dockerfile:
```diff
@@ -0,0 +1,51 @@
+RUN apk del --purge \
+        git \
+    && rm -rf /var/cache/apk/* \
+    && rm -rf $GOPATH \
```
i like this, maybe we should replicate this across the other go services

webstergd commented on src/main/scala/org/holmesprocessing/totem/services/passivetotal2/src/passivetotal-service/main.go:
```diff
@@ -0,0 +1,277 @@
+    // setup http handlers
+    router := httprouter.New()
+    router.GET(analysisURL, handlerAnalyze)
+    router.GET(cfg.Settings.InfoURL, handlerInfo)
```
Not doing a check for InfoURL like you are for analysisURL. However, I would like that eventually to not be a customization option. So ignoring.

ms-xy replied:
That is because it is not necessary for the info url, we don't have a parameter there, we need the slash to make the parameter work correctly

webstergd commented:
last few things and i think this is ready:
- [x] change folder name to passivetotal
- [x] explain in your readme how I can task a domain, ip, or email. 
- [x] run the code though go fmt

ms-xy committed:
Applied go-fmt.

+ Some code rewriting to keep code cleanly formatted even with go-fmt

Signed-off-by: Maximilian Schott <ms@roottec.com>

ms-xy committed:
Renamed directory.

Signed-off-by: Maximilian Schott <ms@roottec.com>

ms-xy committed:
Improved Readme as requested.

Signed-off-by: Maximilian Schott <ms@roottec.com>

ms-xy commented:
The only file that I did not go-fmt is the config file. This is because it will be removed in a pull request soon anyways (will be creating a golang config parsing equivalent in the holmes-library, plus the switch to json config files, which makes this ini config parser obsolete anyways).

webstergd commented:
Thanks ms-xy. This looks good to me. 

I like your identification and validation methods for the ip, domain, and email address. We should eventually move this to the library and make this standard.
